File: DynamicArray.cpp
#include "DynamicArray.h"


template<class T>
DynamicArray<T>::DynamicArray()
{
	flag = true;
	size = 0;
}

template<class T>
void DynamicArray<T>::insert(T data)
{
	this->resize();

	if(flag)
	{
		switch(size)
		{
			case 1:
				array[0] = data;
			break;

			default:
				int index = 0;
				for (; index < size - 1 && array[index] < data ; index++);

				if (array[index] != -1)
				{
					for ( int count = size - 1; count > index ; array[count] = array[count - 1], count--);
					array[index] = data;
				}
				else
				{
					array[index] = data;
				}
			break;
		}
	}
	else
	{
		switch(size)
		{
			case 1:
				array[0] = data;
			break;

			default:
				int index = 0;
				for (; index < size - 1 && array[index] > data ; index++);
					
				if (array[index] != -1)
				{
					for ( int count = size - 1; count > index ; array[count] = array[count - 1], count--);
					array[index] = data;
				}
				else
				{
					array[index] = data;
				}
			break;
		}
	}
}


template<class T>
ostream& operator<<(ostream &os, DynamicArray<T> &obj)
{
	for(int index = 0; index < obj.size; index++)
	{
		if(index + 1 == obj.size)
		{
			os << obj.array[index];
		}
		else os << obj.array[index]<<",";
	}
	return os;
}

template<class T>
void DynamicArray<T>::reverse()
{
	flag = !flag;

	for ( int i = 0, j = size-1 ; i < (size/2); i++,j--)
	{
		T tmp = array[i];
		array[i] = array[j];
		array[j] = tmp;
	}
}

template<class T>
void DynamicArray<T>::resize()
{
	size++;
	T* tmp = new T[size];

	for (int index = 0; index < size-1; index++)
	{
		tmp[index] = array[index];
	}

	array = new T[size];

	for (int index =0; index <size-1; index++)
	{
		array[index] = tmp[index];
	}
	array[size-1] = -1;
}

File: DynamicArray.java
// Creating Generic Dynamic Array

@SuppressWarnings({"unchecked","rawtypes"})
public class DynamicArray<T extends Comparable<T>>
{
	private Comparable[] array;
	private Boolean flag; // if flag is true mean  array store data in ascending order 

	public DynamicArray()
	{
		array = new Comparable[0];
		flag = true;
	}

	public void insert(T data)
	{
		this.resize();
		if (flag == true)
		{
			switch(array.length)
			{
				case 1:
					array[0] = data;
				break;
				default:
					int index = 0;
					for (; index <array.length-1 && array[index].compareTo(data) < 0; index++);
					
					if (array[index] != null)
					{
						for ( int count = array.length - 1; count > index ; array[count] = array[count - 1], count--);
						array[index] = data;
					}
					else
					{
						array[index] = data;
					}
				break;
			}

		}
		else
		{
			switch(array.length)
			{
				case 1:
					array[0] = data;
				break;

				default:
					int index = 0;
					for (; index <array.length-1 && array[index].compareTo(data) > 0; index++);

					if (array[index] != null)
					{
						for ( int count = array.length - 1; count > index ; array[count] = array[count - 1], count--);
						array[index] = data;
					}
					else
					{
						array[index] = data;
					}

				break;
			}
		}
	}

	@Override
	public String toString()
	{
		String str = "[";

		for (int index = 0; index <array.length ; index++)
		{
			if (index + 1 == array.length)
			{
				str +=array[index];
			}
			else str+=array[index]+",";
		}
		str+="]";
		return str;
	}

	public void reverse()
	{
		flag = !flag;

		for ( int i = 0, j = array.length-1 ; i < (array.length/2); i++,j--)
		{
			Comparable tmp = array[i];
			array[i] = array[j];
			array[j] = tmp;
		}
	}

	public void resize()
	{
		Comparable[]  tmp = new Comparable[array.length+1];

		for (int index = 0; index < array.length; index++)
		{
			tmp[index] = array[index];
		}

		array = new Comparable[tmp.length];

		for (int index =0; index <tmp.length-1; index++)
		{
			array[index] = tmp[index];
		}
		array[array.length-1] = null;
	}

}
